[WIP] Validation rules can read coerced argument values (#3873)
graphql-dotnet/graphql-dotnet: docs2/site/docs/migrations/migration8.md

implement your own attributes as shown above, or call the `Validate` method to set a validation
function.

### 6. Validation rules can read or validate field arguments and directive arguments

> Validation rules can now read or validate field arguments and directive arguments. This is useful
> for edge cases, such as when a complexity analyzer needs to read the value of a field argument
> to determine the complexity of the field. While it could also be used within custom validation rules
> to validate the value of field arguments or directive arguments, perhaps based on schema directives,
> it cannot be used to validate fields of input object graph types.

Reviewer: Some parts of the text look like a formal math definition. I had to slow down and read it again. For example, this is too verbose
> Validation rules can now read or validate field arguments and directive arguments.

IMO, this is easier to read
> Validation rules can now read or validate field and directive arguments.

And this happens in many places along the text

```suggestion
Validation rules can now read or validate field and directive arguments. This is useful
for edge cases, such as when a complexity analyzer needs to read the value of a field argument
to determine the complexity of the field. While it could also be used within custom validation rules
to validate the value of field or directive arguments, perhaps based on schema directives,
it cannot be used to validate fields of input objects.
```

Note that with the addition of `Validator` and `Parser` delegates, validation rules are no longer
required to validate input values. If you plan to store validation rules within schema directives,
such that they may be read by clients (see `ExperimentalFeatures.AppliedDirectives`), you may wish
to consider adding a schema initialization visitor to walk the schema, applying the validation rules
to the appropriate fields arguments via the `Validate` method, rather than using a validation rule
to validate the field arguments. This will be much more efficient, as the validation rule will not
be called for every field argument on every field, nor will it require a dictionary lookup to find

> the validation rule for the field argument.

Reviewer: Add a note that both approach example is provided below, so the reader won't waste time trying to understand all the details.

Author: Done

> To implement a validation rule that validates field arguments or directive arguments, implement
> `IValidationRule` and `IVariableVisitorProvider`. The `IVariableVisitorProvider` interface has a
> new method `ValidateArgumentsAsync` that is called during validation. This method is called after
> all field arguments and directive arguments have been parsed, allowing a validation rule to validate
> parsed argument values. It will run even if there are no arguments to validate, so you should check
> for this case and return early if there are no arguments to validate. However it will not run if
> errors occur prior to parsing the arguments, such as if the field or directive is not found, or
> if an argument fails coercion or validation.

Reviewer: ```suggestion
To implement a validation rule that validates field or directive arguments, implement
`IValidationRule` and `IVariableVisitorProvider`. The `IVariableVisitorProvider` interface has a
new method, `ValidateArgumentsAsync`, that is called during validation. This method is called after
all the relevant arguments have been parsed, allowing a validation rule to validate
parsed values. It will run even if there are no arguments to validate, so you should check
for this case and return early if there is nothing to validate. However, it will not run if
errors occur prior to parsing the arguments, such as if the field or directive is not found, or
if an argument fails coercion or validation.
```

        var fieldDef = context.TypeInfo.GetFieldDef();
        if (fieldDef == null)
            return default;
        // look for connection types
        if (fieldDef.ResolvedType?.GetNamedType() is not IObjectGraphType connectionType || !connectionType.Name.EndsWith("Connection"))
            return default;
        // retrieve the arguments
        if (!(context.ArgumentValues?.TryGetValue(fieldNode, out var args) ?? false))
            return default;
        // look for first and last arguments
        ArgumentValue lastArg = default;
        if (!args.TryGetValue("first", out var firstArg) && !args.TryGetValue("last", out lastArg))
            return default;
        var rows = (int?)firstArg.Value ?? (int?)lastArg.Value ?? 0;
        if (rows > 1000)
            context.ReportError(new ValidationError("Cannot return more than 1000 rows"));
        return default;
    }

    ValueTask INodeVisitor.LeaveAsync(ASTNode node, ValidationContext context) => default;
}
```

The above is sample functionality, although in this situation it would be more efficient to
have a schema visitor apply this validation rule to the appropriate fields via the `Validate`

> method during schema initialization.

Reviewer: Can you add an example here to clarify the difference between the two approaches?

Author: Done

if you wish to define your own naming logic.

The `UseLegacyTypeNaming` option is deprecated and will be removed in GraphQL.NET v9.

> ### 9. `IVariableVisitorProvider.ValidateArgumentAsync` added

Author: ```suggestion
### 9. `IVariableVisitorProvider.ValidateArgumentsAsync` added
```

> The `IVariableVisitorProvider` interface has a new method `ValidateArgumentAsync` that is

Author: ```suggestion
The `IVariableVisitorProvider` interface has a new method `ValidateArgumentsAsync` that is
```

For federated schemas, the `ServiceGraphType`'s `sdl` field will now use the
new implementation to print the schema. Please raise an issue if this causes
a problem for your federated schema.

### 11. `IVariableVisitorProvider.ValidateArgumentsAsync` added

> The `IVariableVisitorProvider` interface has a new method `ValidateArgumentsAsync` that is
> called during validation. This method is called after all field arguments and directive
> arguments have been parsed, allowing a validation rule to validate parsed argument values.

Reviewer: ```suggestion
The `IVariableVisitorProvider` interface has a new method, `ValidateArgumentsAsync`, that is
called during validation. This method is called after all field and directive
arguments have been parsed, allowing a validation rule to validate parsed argument values.
```
